File: 백준/Gold/17298. 오큰수/오큰수.py
# 1초, 512mb
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
nge = [-1 for _ in range(n)]
stack = [0] # 인덱스 저장할 스택

for i in range(1, n):
    while stack and arr[stack[-1]] < arr[i]:
        nge[stack[-1]] = arr[i]
        stack.pop()
    stack.append(i)
    
print(*nge)

File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.*;
import java.util.*;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());

        int[] numbers = new int[N];
        for(int i=0; i<N; i++) {
            numbers[i] = Integer.parseInt(st.nextToken());
        }

        Stack<Integer> stack = new Stack<>();
        stack.push(0);

        int[] answer = new int[N];
        for(int i=1; i<N; i++) {
            while(!stack.isEmpty() && numbers[stack.peek()] < numbers[i]) {
                answer[stack.pop()] = numbers[i];
            }
            stack.push(i);
        }

        while(!stack.isEmpty()) {
            answer[stack.pop()] = -1;
        }

        StringBuilder sb = new StringBuilder();
        for(int i: answer) {
            sb.append(i).append(" ");
        }
        System.out.println(sb);
    }
}